### ISSUE TITLE
Linux(WSL Ubuntu) installation curl command fails

### ISSUE THREAD

**Issue by UeberTimei** [ollama/ollama#2448]
curl -fsSL https://ollama.com/install.sh | sh

This leads to:
curl: (35) OpenSSL SSL_connect: Connection reset by peer in connection to ollama.com:443

I tried everything. I reinstalled WSL and set Google DNS.

**Comment by amnweb** [comment]
Did you try to run `sudo update-ca-certificates` ?

And what's the output for `curl -v http://ollama.com/install.sh` ?

**Comment by pdevine** [comment]
I think there was an issue w/ this when we switched from `ollama.ai` to `ollama.com`. Can you try it with:

```
curl -fsSL https://ollama.ai/install.sh | sh
```

**Comment by UeberTimei** [comment]
I've installed ollama on WSL but now I can't install models.

pulling manifest
Error: pull model manifest: Get "https://registry.ollama.ai/v2/library/llama2/manifests/latest": dial tcp: lookup registry.ollama.ai on [::1]:53: read udp [::1]:46273->[::1]:53: read: connection refused

**Comment by dhiltgen** [comment]
Are you perhaps behind a firewall or proxy?

Can you try to curl the site from inside a wsl terminal?  Something like `curl https://registry.ollama.ai/v2/_catalog` (takes a while, but should produce a json blob showing the first few dozen public models)

Perhaps our Windows native install is an option to consider?

**Comment by jianotl** [comment]
root@f09b3869a542:/# ollama run gemma
pulling manifest 
Error: pull model manifest: Get "https://ollama.com/token?nonce=7zKHsvSwLKftB_ZFljykiw&scope=repository%!A(MISSING)library%!F(MISSING)gemma%!A(MISSING)pull&service=ollama.com&ts=1710593137": read tcp 172.17.0.6:33406->34.120.132.20:443: read: connection reset by peer

i also get it

**Comment by dhiltgen** [comment]
@jianotl can you try the `curl` test I suggested above?  My suspicion is there's a firewall or proxy configuration that's preventing network access to the ollama model registry.

**Comment by mxyng** [comment]
@UeberTimei port 53 indicates there's a problem with DNS. Make sure DNS is set up correctly in this environment

**Comment by UeberTimei** [comment]
> Are you perhaps behind a firewall or proxy?
> 
> Can you try to curl the site from inside a wsl terminal? Something like `curl https://registry.ollama.ai/v2/_catalog` (takes a while, but should produce a json blob showing the first few dozen public models)
> 
> Perhaps our Windows native install is an option to consider?

I'm getting this:
curl: (6) Could not resolve host: registry.ollama.ai

**Comment by mxyng** [comment]
> curl: (6) Could not resolve host: registry.ollama.ai

This indicates there's a problem with DNS. Please ensure the DNS is set up correctly for the WSL context

**Comment by dhiltgen** [comment]
@UeberTimei I think this most likely indicates a local network, vpn or proxy problem.  The site is online and accessible.

If you're running on Windows, you might consider the native Windows install of Ollama if you're having difficulty setting up WSL2 networking to work correctly in your environment.

**Comment by chopeen** [comment]
I am also experiencing the same problem:

```
$ ollama run llava
pulling manifest 
Error: pull model manifest: Get "https://ollama.com/token?nonce=SW96RgmctcQmHJ37NXJ8KQ&scope=repository%!A(MISSING)library%!F(MISSING)llava%!A(MISSING)pull&service=ollama.com&ts=1711106784": read tcp 10.144.68.189:40600->34.120.132.20:443: read: connection reset by peer
```

Downloading the catalog is working fine:

```
$ curl https://registry.ollama.ai/v2/_catalog
{"repositories":["a123/tiny-gpt","abambah/avnit","abhi3linku/test-model","adeelahmad/aws-zephyr-7b-alpha","adrienbrault/biomistral-7b","adrienbrault/gorilla-openfunctions-v2","adrienbrault/lonestriker-senku-70b","adrienbrault/nous-capybara-3b","adrienbrault/nous-hermes2pro","adrienbrault/qwen1.5-0.5b-openhermes-2.5","adrienbrault/saul-instruct-v1","adrienbrault/tess-70b-v1.6","adrienbrault/test-phi2","adrienbrault/test-q1.5-0.5b","adrienbrault/test-tv1b","adrienbrault/testtd","adrienbrault/wolfram-miquliz-120b-v2","adrienbrault/yi-9b-200k","adrienlcq/shad0w59i","aimadeapproachable/genstruct_7b","airat/karen-the-editor-v2-creative","airat/karen-the-editor-v2-strict","aish/svce-ai","aisherpa/codellama-70b-instruct-hf","aisherpa/mistral-7b-instruct-v02","aisherpa/wizardmath-7b-v1.1","alenbijelic/devops","alhu/zephyr-chat","ancerlop/mistral-params","andrewcanis/command-r","anik/copilot_llama2","ankk98/cyrax7b","antony66/saiga_mistral_7b_128k","anxiu0101/campus-hub-bot","apattana/llama2","apattana/new-model","apattana/test-model","apto/elyza","argilla/notus","ariekdev/oy_bro","arsalananwar/academate","asedmammad/contextual-mistral","asedmammad/lhk-dpo","avinish/finetuned-qunatized","azure99/blossom-test","azure99/blossom-v5","bbrittuw/bge-large-en-v1.5","bbrittuw/ggml-sfr-embedding-mistral-q8_0","bdx0/vietcuna","bengt0/em_german_leo_mistral","bhargav/donkey-ai","bhargav/donkey-initial","bhargav/donkey-soql","bigllama/mistralv01-7b","bitbinge/dazz","bitbinge/stickler","bitbinge/zoar","bjaburg/gen4-mistral","bjaburg/porsche-phi","bob/gemma-test-1","boubou77/emojicool","boubou77/zeus","breitburg/tinyllama","bruceignoretest/bruce-test-1","bruceignoretest/bruce-test-2","bruceignoretest/bruce-test-3","bruceignoretest/bruce-test-4","brunneis/hermes-2-pro-mistral-7b-q8-0","brunneis/hermes-2-pro-mistral-7b","brunneis/mistral-7b-instruct-v0.2-q4-k-m","brxce/gemma","brxce/mario","brxce/mario23","brxce/marioo","brxce/monadgpt","brxce/ne","brxce/stable-diffusion-prompt-generator","brxce/test-fix","brxce/whiterabbitneo-33b","c2p/gemma-ai","cagataycali/neuron","calebfahlgren/natural-functions","captainkyd/trinity-13b","captainkyd/whiterabbitneo7b","carotamunoz/mario","cas/alphamonarch-7b","cas/brezn-7b","cas/brezn3","cas/discolm-german-laser","cas/discolm-mfto-german","cas/german-assistant-v7","cas/hermeo-7b","cas/kafkalm-7b-dare-ties-laserrmt-qlora-dpo-v0.5","cas/kafkalm-7b-german-v0.1","cas/minicpm-3b-hephaestus","cas/minicpm-3b-openhermes-2.5-v2","cas/minicpm-3b-turangga-v3-ep50","cas/mistral-ft-optimized-1227","cas/mistral-instruct-v0.2-2x7b-moe","cas/mixtral_11bx2_moe"]}
```

Port `443` suggests an SSL issue and I have some company self-signed certificates added to `/etc/ssl/certs/ca-certificates.crt`, so that the OS trusts them.

1. Is it possible that the Ollama application rejects them nonetheless?
2. Is it possible to run Ollama in verbose mode or check details in any log file?

**Comment by dhiltgen** [comment]
> Is it possible that the Ollama application rejects them (self signed proxy certs) nonetheless?

This sounds like a plausible explanation.  That said, it sounds like you updated the expected file for ubuntu.

https://go.dev/src/crypto/x509/root_linux.go

You might want to try a few other paths from that file and see if one works.  If not, it may require code changes to ollama to be able to adjust how we're establishing SSL connections.

> Is it possible to run Ollama in verbose mode or check details in any log file?

We do have `OLLAMA_DEBUG=1` for turning up verbosity, but I don't think that will yield information about TLS trusted root cert lookups.

**Comment by chopeen** [comment]
I created a [simple Go application](https://gist.github.com/chopeen/8b2943db321d25d338e20f6283430a2e) to test downloading file - all the URLs I checked were fetched fine (sometimes the content is empty, but no SSL errors).

I think the error occurs in function `PullModel`, but I don't know how to call it manually (I'm Python developer with zero Go experience).

I was able to run the unit tests - they're all green except for `integration`:

![image](https://github.com/ollama/ollama/assets/183731/63b41197-f495-49ff-b150-c8104e56c475)

Should I try and run the integration tests? Do they pull actual models from the Ollama library?

**Comment by ccsyt1205** [comment]
@dhiltgen Sorry to disturb you bro, I encountered a problem while executing 'ollama pull gemma' behind a  proxy, is there any solution to specify a proxy? Maybe like this: ollama pull gemma --proxy "http:xxxxxx.com.cn"?

**Comment by chopeen** [comment]
@ccsyt1205 I think you are looking for `HTTP_PROXY` or `HTTPS_PROXY` - see [faq.md#how-do-i-use-ollama-behind-a-proxy](https://github.com/ollama/ollama/blob/main/docs/faq.md#how-do-i-use-ollama-behind-a-proxy) for details.

**Comment by chopeen** [comment]
I realized that Ollama used to work fine for me in the office, so I **downgraded** Ollama to v0.1.24 and the problem is **gone**. 🎉

I don't know yet if that is the most recent version that will work at the office, but something must have changed when it comes to network connections. I will go through the change log and try to pinpoint the exact change.

**Comment by mxyng** [comment]
#2719 is unrelated. Furthermore, this issue has gotten off topic. The original issue deals with fetching the install script which has been resolved. I'm now going to close this issue.

For further discussion on pull errors, specifically connections being forcefully closed, please see #3112. Other similar issues will be closed as duplicates
